### ISSUE TITLE
Feat: Add field sensitivity and other metadata information to state file resource attributes.

### ISSUE THREAD

**Issue by mhaley-miovision** [hashicorp/terraform#25646]
### Current Terraform Version

```
Terraform v0.12.28
```

### Use-cases

Determine if a resource attribute is sensitive or not by using only the statefile.  This sort of information is useful for 3rd party tools such as terraboard so that they can correctly blank out sensitive attributes of resources before displaying them to the user.

### Attempted Solutions

I have used the `random_password` resource that sets it's result to sensitive (defined [here](https://github.com/hashicorp/terraform-provider-random/blob/master/random/resource_password.go) and ultimately set [here](https://github.com/hashicorp/terraform-provider-random/blob/3604672aecc38375724552d1b6d12036d5a33d1a/random/string.go#L94)) and then viewed the statefile.  The corresponding resource block looks like this
```
{
      "mode": "managed",
      "type": "random_password",
      "name": "password",
      "provider": "provider.random",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "none",
            "keepers": null,
            "length": 16,
            "lower": true,
            "min_lower": 0,
            "min_numeric": 0,
            "min_special": 0,
            "min_upper": 0,
            "number": true,
            "override_special": "_%@",
            "result": "<random-password-here>",
            "special": true,
            "upper": true
          },
          "private": "bnVsbA=="
        }
      ]
    }
``` 

As you can see, I have no easy way of knowing that the result is in fact marked as sensitive just from looking at the statefile.

### Proposal

From the statefile perspective we could make each attribute of a resource an object that contains the actual value as well as any other meta data about that attribute such as type.  For example

```
  {
      "mode": "managed",
      "type": "random_password",
      "name": "password",
      "provider": "provider.random",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "result": {
              "value": "<random-password-here>",
              "sensitive": true,
              "type": "string"
            }
          },
          "private": "bnVsbA=="
        }
      ]
    }
```

**Comment by apparentlymart** [comment]
Hi @mhaley-miovision,

Under the current design, we expect systems that need information like this to be using the `terraform show -json` command in conjunction with the `terraform providers schema -json` command, the former because it is designed for external consumption rather than Terraform's own consumption and the latter because it gives the necessary metadata to fully interpret the information in the state.

For example, Terraform Cloud and Terraform Enterprise follow this strategy with saved plans in order to expose the relevant data to the Sentinel policy engine, and so when Terraform Cloud creates a plan it also runs `terraform show -json planfile` and `terraform providers schema -json` to capture that metadata for use by the policy enforcement step.

The main Terraform state snapshots are not intended for external consumption, and are instead optimized for what Terraform itself needs. Terraform already has access to the provider schema metadata, so it doesn't redundantly store that metadata in the state snapshots. Storing schema metadata inline for every attribute would, I think, make the resulting snapshots significantly larger and thus more costly to transfer and store, and potentially more likely to hit storage size limits in some backends.

With all of that said, I realize that Terraboard's current design doesn't lend itself well to this intent, because it's expecting to just be able to pull a state snapshot directly from the storage backend and handle it with no additional information. That isn't a use-case Terraform is intending to support today. I'm going to consider this a feature request to support that use-case somehow, but I want to be explicit that adding redundant additional information to the state snapshot format is likely not the preferred way to meet that use-case due to the tradeoffs I described above.

In the meantime, I think any system that intends to produce a visualization of the state will either need to accept that it can't reliably detect sensitive values :confounded: or will need to be integrated with a broader Terraform automation solution like Terraform Cloud, so that it can capture the information it needs in the supported way as part of the process of applying a change and store it separately from the main state snapshot artifact.

**Comment by mhaley-miovision** [comment]
Those are all valid points.  I think that asking 3rd party applications to run the state file through an extra process isn't too much to ask.  In the case of terraboard, it might take a bit of extra work, but it is doable. 

The problem I am struggling with right now in a quick example that I spun up is that terraform heavily relies on a pre-existing terraform cache in order to parse the tfstate.json file.  In my test I took a tfstate file and copied it into a new directory.  I then run `terraform show tfstate.json` and it complains that it is missing schemas for the different providers.  
```
# data.aws_subnet_ids.private:
# missing schema for provider "aws"
```
I tried to copy providers from a working terraform directory's cache `.terraform/plugin/linux_amd64/terraform-provider-aws_v2.70.0_x4` into `~/.terraform.d/plugins/`, and even tried to rename the provider to `terraform-provider-aws` with no luck.  So as it stands today, this workflow is still very unusable for 3rd party applications.

In our case we are big users of terragrunt too, so we really don't have a main.tf that we could point an application to even if we wanted to try and make this pipeline work.

**Comment by apparentlymart** [comment]
Indeed, from what I was able to quickly understand of the design of Terraboard I expect that probably the most practical design would be to allow the user to submit the output of running `terraform providers schema -json` as part of the configuration of a particular workspace (or whatever noun Terraboard uses for that) and then, when rendering the state, be tolerant of the submitted schema potentially being incomplete (e.g. because the user has added a new provider to the configuration since they originally set it up), prompting the user to update the schema to get a complete result.

I'm not pretending that this is an _elegant_ solution, but rather a compromise relying on the fact that the set of providers in a configuration tends to change much less frequently than the _state_ associated with a configuration, and so I would expect that providing the schema out of band would not be _too_ arduous.

The commands I mentioned are intended to be run in an already-initialized directory, often immediately after completing a `terraform init` (for schema), `terraform plan` (for `terraform show -json PLANFILE`), or `terraform apply` (for `terraform show -json`), and so indeed they are not useful for a program that is not directly involved in the main workflow, like Terraboard. They were designed to be used by the main orchestration tool itself, to be run in the same execution environment as the command that generated the artifact they are inspecting.

The requirement to have the providers installed to get the schema is not really negotiable since the providers are where the schema comes from, but we could potentially introduce other ways to grab the schema for a particular provider version, independent of any particular Terraform configuration, to better support an approach like what you tried there, though I would expect that having Terraboard effectively retrieve and execute arbitrary code in order to execute the schema would not be desirable: it would force Terraboard to have an isolated execution environment just like a typical Terraform automation wrapper does.

**Comment by mhaley-miovision** [comment]
I was not aware of the `terraform providers schema -json` command.  If that is doing what I think it does, could we not have the `terraform show` command have an option to pass in the `schema.json` so that it can parse the tfstate file?

>The requirement to have the providers installed to get the schema is not really negotiable since the providers are where the schema comes from, 

Does the `terraform providers schema -json` command not provide this information?  I guess the schema.json would need to be versioned too to be able to handle different version of a tfstate file.  How does terraform handle different state file versions internally?  If I had a state file that used a very old version of a provider, how does terraform parse it?  Does it just download an older version of the provider and that's it?

> I would expect that having Terraboard effectively retrieve and execute arbitrary code in order to execute the schema would not be desirable: it would force Terraboard to have an isolated execution environment just like a typical Terraform automation wrapper does.

I'm not sure if this has to be the case.  If provider information were stored in the state file, then `terraform show` could read that provider information, download the needed providers, and then parse/show the state.  I don't know if an isolated environment is needed?  We certainly don't use isolated environments when we are applying the actual tf scripts against our accounts.  We just run `terraform apply` from our laptops.  Unless terraform itself provides this isolated environment that you are referring to?

**Comment by apparentlymart** [comment]
The `terraform providers schema -json` output is generated _from_ the schema stored in the providers. Terraform does not currently have any situation where it parses what it produced from that, because Terraform just asks the provider directly for schema. You're right that in theory the `terraform show` command could have an option to read the schema from a JSON file produced earlier, and that could be a way to address this use-case if you were to establish some way for Terraboard to already have a copy of the schema JSON file.

To answer your question about provider schema versions: the `schema_version` property we can see in your original example is an integer chosen by the provider to mark a particular version of the schema for that resource type. When a provider introduces a new version of schema, we expect it to also provide logic to upgrade from the previous version to that version. As part of reading the state in preparation for other operations, Terraform Core asks the provider to upgrade the stored state to be compatible with whatever schema version is current for the installed version of the provider. Until that upgrade step is complete, Terraform Core just handles the value of the `attributes` property as an opaque JSON string and therefore doesn't need access to the earlier schema versions.

**Comment by pselle** [comment]
Without diving too deeply into the discussion on this post, I believe this is solved by the recent addition of sensitivity-tracking through input variables and sensitive output variables in Terraform 0.14 (now in beta). 

As a function of keeping track of sensitive paths so that updates and destroys aware of sensitivity, the statefile now includes a field for resources `sensitive_attributes` that keeps track of paths within the resource that should be considered sensitive. 

I'm adding this issue to the 0.14 milestone with this in mind, please chime in if the extended discussion diverged far enough from the original issue that my assertion here is not true.

**Comment by mhaley-miovision** [comment]
Yeah, it sounds like that would get the desired information to be available within the tfstate file.  The underlying goal here is to allow 3rd party apps like terraboard to be able to look only at the tfstate file to and hide any values that are sensitive from users.

**Comment by pselle** [comment]
Defining input variables as sensitive in order to redact the values from UI output is now available in 0.14! [A recent blog post](https://www.hashicorp.com/blog/terraform-0-14-adds-the-ability-to-redact-sensitive-values-in-console-output). From the request here, the information to target sensitive paths within resources is now available in the statefile.
